### PR TITLE
improve options parsing

### PR DESCRIPTION
**Pull request by snoyer**
First pass on the new options IO. Here we improve parsing for the existing `bool`, `int`, `double`, `ratio_t` types.

I also split the IO tests from the options structure tests (the options tests could possibly be reduced in the future as they currently also do some parsing/formatting test).

### CODE REVIEW

File: library/private/options_tools.h.in [hunk 1]
```diff
@@ -23,6 +24,32 @@ std::string trim(std::string_view strv)
   return std::string(strv);
 }
 
+//----------------------------------------------------------------------------
+/** Call `std::stod` and throw `std::invalid_argument` if the whole string has not been consumed. */
+double parse_double(const std::string& str)
```
Comment (snoyer) [inline library/private/options_tools.h.in:29]:
```suggestion
double stodStrict(const std::string& str)
```
better name? not as readable but more explicit

Comment (mwestphal) [reply library/private/options_tools.h.in:29]: not a big fan of this tbh, I like the symmetry of having one method by type, named logically

File: library/private/options_tools.h.in [hunk 2]
```diff
@@ -23,6 +24,32 @@ std::string trim(std::string_view strv)
...
+{
+  std::size_t pos = 0;
+  const double d = std::stod(str, &pos);
```
Comment (mwestphal) [inline library/private/options_tools.h.in:32]: not a big fan of single letter var

File: library/private/options_tools.h.in [hunk 3]
```diff
@@ -23,6 +24,32 @@ std::string trim(std::string_view strv)
...
+  }
+  return d;
+}
+
+//----------------------------------------------------------------------------
+/** Call `std::stoi` and throw `std::invalid_argument` if the whole string has not been consumed */
+int parse_int(const std::string& str)
+{
+  std::size_t pos = 0;
+  const int i = std::stoi(str, &pos);
```
Comment (mwestphal) [inline library/private/options_tools.h.in:45]: not a big fan of single letter var

File: library/private/options_tools.h.in [hunk 4]
```diff
@@ -23,6 +24,32 @@ std::string trim(std::string_view strv)
...
+  if (pos != str.size())
+  {
+    throw std::invalid_argument("partial");
+  }
+  return i;
+}
+
 //----------------------------------------------------------------------------
 /**
  * Vector specific templated parse method
```
Comment (mwestphal) [inline library/private/options_tools.h.in:55]: mention the trimming ?

File: library/private/options_tools.h.in [hunk 5]
```diff
@@ -78,7 +110,7 @@ int parse(const std::string& str)
   // TODO implement more parsing possibilities ?
```
Comment (mwestphal) [inline library/private/options_tools.h.in:110]: remove this comment  ?

File: library/private/options_tools.h.in [hunk 6]
```diff
@@ -103,7 +135,7 @@ double parse(const std::string& str)
...
```
Comment (mwestphal) [inline library/private/options_tools.h.in:135]: remove the comment ?

File: library/private/options_tools.h.in [hunk 7]
```diff
@@ -45,25 +72,30 @@ T parse(const std::string& str)
...
+    vec.emplace_back(options_tools::parse<typename T::value_type>(options_tools::trim(each)));
   }
   return vec;
 }
 
 //----------------------------------------------------------------------------
 /**
  * Parse provided string into a bool
- * Numerical and boolaplha are both supported
+ * Supported values are: `"true"`/`"false"`, `"1"`/`"0"`, `"yes"`/`"no"`, `"on"`/`"off"`
```
Comment (mwestphal) [inline library/private/options_tools.h.in:83]: mention the lowering ?

File: library/private/options_tools.h.in [hunk 8]
```diff
@@ -125,8 +157,23 @@ double parse(const std::string& str)
...
 {
-  // TODO implement proper ratio parsing
-  return options_tools::parse<double>(str);
+  try
+  {
+    if (!str.empty() && str.at(str.size() - 1) == '%')
+    {
+      return parse_double(str.substr(0, str.size() - 1)) / 100;
+    }
+    return parse_double(str);
```
Comment (mwestphal) [inline library/private/options_tools.h.in:166]: supports fraction ? 

eg: `1/3` ?

File: library/private/options_tools.h.in (hunk 1 above)
Comment (snoyer) [reply library/private/options_tools.h.in:29]: the template version throws "user friendly" exception mentioning the type name, so if we can't really use `double parse()` from within `ratio_t parse()`

File: library/private/options_tools.h.in [hunk 9]
```diff
@@ -125,8 +157,23 @@ double parse(const std::string& str)
 template<>
 ratio_t parse(const std::string& str)
 {
-  // TODO implement proper ratio parsing
-  return options_tools::parse<double>(str);
+  try
+  {
+    if (!str.empty() && str.at(str.size() - 1) == '%')
```
[the remark this answers is not in the document]
Comment (snoyer) [reply library/private/options_tools.h.in:162]: possible, would have done it already if it was ascii but encoding is scary, would rather avoid :)

File: library/private/options_tools.h.in (hunk 8 above)
Comment (snoyer) [reply library/private/options_tools.h.in:166]: slippery slope to full blown calculator but if we limit to fractions i could see it

File: library/private/options_tools.h.in (hunk 9 above)
[the remark this answers is not in the document]
Comment (mwestphal) [reply library/private/options_tools.h.in:162]: indeed, fair

File: library/private/options_tools.h.in [hunk 10]
```diff
@@ -103,7 +135,7 @@ double parse(const std::string& str)
   // TODO implement more parsing possibilities ?
   try
   {
-    return std::stod(str);
+    return parse_double(str);
```
Comment (mwestphal) [inline library/private/options_tools.h.in:138]: supports NaN and Inf ?

File: library/private/options_tools.h.in [hunk 11]
```diff
@@ -78,7 +110,7 @@ int parse(const std::string& str)
   // TODO implement more parsing possibilities ?
   try
   {
-    return std::stoi(str);
+    return parse_int(str);
```
Comment (mwestphal) [inline library/private/options_tools.h.in:113]: supports NaN and Inf ?

File: library/private/options_tools.h.in (hunk 10 above)
Comment (snoyer) [reply library/private/options_tools.h.in:138]: should be possible

File: library/private/options_tools.h.in (hunk 11 above)
Comment (snoyer) [reply library/private/options_tools.h.in:113]: there's no way to properly encode that as `int` tho, is there? I guess would could fake `±inf` with some kind of `MIN_INT`/`MAX_INT` but is it worth it?

File: library/private/options_tools.h.in (hunk 1 above)
Comment (mwestphal) [reply library/private/options_tools.h.in:29]: got it, in that case I prefer `stodStrict` but maybe @Meakk have inputs here.

Comment (michaelmigliore) [reply library/private/options_tools.h.in:29]: `stodStrict` or `stodThrow` or `stodSafe` are all fine

Comment (Meakk) [reply library/private/options_tools.h.in:29]: `stodStrict` or `stodThrow` or `stodSafe` are all fine

File: library/private/options_tools.h.in (hunk 8 above)
Comment (Meakk) [reply library/private/options_tools.h.in:166]: I think it's useful too. Much easier to set `1/30` when setting the framerate than `0.0333333333333333`.

File: library/private/options_tools.h.in (hunk 10 above)
Comment (snoyer) [reply library/private/options_tools.h.in]: possible but current use-cases would be better implemented with proper optional value support

File: library/private/options_tools.h.in (hunk 11 above)
Comment (snoyer) [reply library/private/options_tools.h.in]: let's wait until we have a use-case

File: library/private/options_tools.h.in (hunk 10 above)
Comment (mwestphal) [reply library/private/options_tools.h.in]: indeed

File: library/private/options_tools.h.in [hunk 12]
```diff
@@ -23,9 +24,36 @@ std::string trim(std::string_view strv)
   return std::string(strv);
 }
 
+//----------------------------------------------------------------------------
+/** Call `std::stod` and throw `std::invalid_argument` if the whole string has not been consumed. */
```
Comment (mwestphal) [inline library/private/options_tools.h.in:28]: maybe document this in options.h as the exception output will ba a bit non-descriptive from a libf3d user POV

File: library/private/options_tools.h.in [hunk 13]
```diff
@@ -45,25 +73,31 @@ T parse(const std::string& str)
...
   }
   return vec;
 }
 
 //----------------------------------------------------------------------------
 /**
  * Parse provided string into a bool
- * Numerical and boolaplha are both supported
+ * Supported values are: `"true"`/`"false"`, `"1"`/`"0"`, `"yes"`/`"no"`, `"on"`/`"off"`
+ * (case-insensitive)
```
Comment (mwestphal) [inline library/private/options_tools.h.in:85]: update options.h doc too

File: library/private/options_tools.h.in [hunk 14]
```diff
@@ -118,15 +150,37 @@ double parse(const std::string& str)
 
 //----------------------------------------------------------------------------
 /**
- * Parse provided string into a ratio_t
- * Rely on <double> version for parsing
+ * Parse provided string into a ratio_t.
+ * Supported formats: number (same as `<double>` version), percentage, fraction
```
Comment (mwestphal) [inline library/private/options_tools.h.in:154]: give some examples, put doc in options.h too

**Review by mwestphal — CHANGES_REQUESTED** [review]
small comments about doc, maybe the doc should only be in options.h

File: library/private/options_tools.h.in (hunk 12 above)
Comment (snoyer) [reply library/private/options_tools.h.in:28]: `stod` already throws its own `std::invalid_argument` in some cases, we just add an additional one here and it should be caught by the templated `parse` functions and not bubble up to the users.
Also this function doesn't have to be exposed to the outside actually (but I don't know what that implies and how to deal with it)

Comment (mwestphal) [reply library/private/options_tools.h.in:28]: Indeed, ill need to doc that in the .md file though

File: library/private/options_tools.h.in (hunk 14 above)
Comment (mwestphal) [reply library/private/options_tools.h.in:154]: after discussions, no need for examples for now. just put that doc in options.h

**Review by mwestphal — APPROVED** [review]

**Review by Meakk — APPROVED** [review]

**Pull request merged**